### PR TITLE
Fix the logic when setting up the ptree kblk_reader

### PR DESCRIPTION
**Pull request by tristan957**
This part of the code since the hblock was conceived has essentially
been one big hack for better or for worse. There has been a bit of
fallout in this area of the code. See the last two PRs by me.

In this case, we have a few changes that'll hopefully fix the stalling
being seen by Greg during mongo replication testing. `struct
kblk_reader` is a misnomer for the most part. Really it is a reader that
can span multiple WBTrees. In the case of the ptomb kblk_reader, there
is only ever one block to read. The ptomb WBTree is stored entirely in
the kvset's singular hblock. The bounds of kblk_reader::kr_kblk_cnt were
wrong in the case of the ptomb reader.

Signed-off-by: Tristan Partin <tpartin@micron.com>

### CODE REVIEW

File: lib/cn/kvset.c [hunk 1]
```diff
@@ -2390,7 +2388,8 @@ kblk_start_read(struct kvset_iterator *iter, struct kblk_reader *kr, enum read_t
             return;
         }
 
-        iter->curr_kblk = kr->kr_next_kblk_idx;
+        iter->curr_kblk = kr->kr_next_blk_idx;
+        kr->kr_next_blk_idx++;
```
Comment (tristan957) [inline lib/cn/kvset.c:2392]: Note this was broken in master from my previous PR, so it has been restored

File: lib/cn/kvset.c [hunk 2]
```diff
@@ -2358,24 +2358,22 @@ kblk_start_read(struct kvset_iterator *iter, struct kblk_reader *kr, enum read_t
...
+            kblk = &iter->ks->ks_kblks[kr->kr_next_blk_idx];
+
             wbt = &kblk->kb_wbt_desc;
             kr->kr_mbid = kblk->kb_kblk.bk_blkid;
-            kr->kr_next_kblk_idx++;
             break;
         case READ_PT:
             wbt = &iter->ks->ks_hblk.kh_ptree_desc;
             kr->kr_mbid = iter->ks->ks_hblk.kh_hblk.bk_blkid;
-            /* Reuse same kblock for next iteration since it went unused */
```
Comment (tristan957) [inline lib/cn/kvset.c:2378]: @beckerg here you go :)

File: lib/cn/kvset.c [hunk 3]
```diff
@@ -2551,8 +2549,9 @@ kvset_iter_enable_mblock_read_pt(struct kvset_iterator *iter)
     if (ev(err))
         return err;
 
-    /* ptombs are stored in the last kblock */
-    kr->kr_next_kblk_idx = kr->kr_kblk_cnt - 1;
+    /* ptombs are stored in the singular hblock */
+    kr->kr_blk_cnt = 1;
```
Comment (tristan957) [inline lib/cn/kvset.c:2553]: Here was a large issue.

File: lib/cn/kvset.c [hunk 4]
```diff
@@ -2528,7 +2527,6 @@ kvset_iter_enable_mblock_read_cmn(struct kvset_iterator *iter, struct kblk_reade
 
     kr->asyncio = iter->asyncio;
 
-    kr->kr_kblk_cnt = iter->ks->ks_st.kst_kblks;
```
Comment (tristan957) [inline lib/cn/kvset.c:2531]: Initialized elsewhere.

File: lib/cn/kvset.c [hunk 5]
```diff
@@ -2074,13 +2074,13 @@ struct kblk_reader {
     } iores;
 
     u64 kr_mbid;
-    u16 kr_kblk_cnt;
+    u16 kr_blk_cnt;
```
Comment (tristan957) [inline lib/cn/kvset.c:2077]: Renamed to better reflect what it going on.

File: lib/cn/kvset.c [hunk 6]
```diff
@@ -2350,32 +2350,32 @@ static void
...
-    struct kvset_kblk *kblk;
 
     assert(!kr->mbio.pending);
     assert(iter->workq);
 
     if (kr->kr_nodex == kr->kr_nodec) {
         struct wbt_desc *wbt;
 
-        if (kr->kr_next_kblk_idx == kr->kr_kblk_cnt) {
+        if (kr->kr_next_blk_idx == kr->kr_blk_cnt) {
```
Comment (beckerg) [inline lib/cn/kvset.c:2360]: I like to use >= in cases like this to be somewhat more defensive, then you could drop an assert(blk_idx == blk_cnt) inside the if block to be extra cautious...

**Review by beckerg — APPROVED** [review]

**Review by alexttx — APPROVED** [review]

**Pull request merged**